### PR TITLE
Check on active account vs environment (which is shallow copied)

### LINKED ISSUE #367: Can not detect Bec.sol int overflow
**Describe the bug**
Try `./myth -x Bec.sol`.
But no excepted report given, the only thing is the bellow.
I am not sure if I missed some options, can anyone help?

```
==== Exception state ====
Type: Informational
Contract: BecToken
Function name: **ambiguous** t
PC address: 5260
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: solidity_examples/BECToken.sol:28

assert(c >= a)

--------------------
```

### PR DESCRIPTION
**Pull request by JoranHonig**
Fixes #367

### CODE REVIEW

**Comment by rocky** [comment]
Should a test be added to include this? Or is it covered by some existing benchmark that we might hook into in the future?

**Comment by JoranHonig** [comment]
@rocky Unfortunately we can't effectively unit test this as the bug arises from the interplay of several components. I'll add a testcase for the report_test

**Comment by b-mueller** [comment]
I'm still not getting the expected result, only the mathlib ADD overflow is detected.

```
$ ./myth -x solidity_examples/BECToken.sol -minteger
==== Integer Overflow ====
Type: Warning
Contract: BasicToken
Function name: **ambiguous** t
PC address: 993
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Integer Overflow ====
Type: Warning
Contract: StandardToken
Function name: **ambiguous** t
PC address: 2727
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Integer Overflow ====
Type: Warning
Contract: PausableToken
Function name: **ambiguous** t
PC address: 4216
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: **ambiguous** t
PC address: 5247
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------
```

Gonna try to re-run with higher max-depth/timeout.

**Comment by b-mueller** [comment]
Similar result for max_depth=32, timeout = 20,000 (although additional exceptions are found).

```
==== Integer Overflow ====
Type: Warning
Contract: BasicToken
Function name: **ambiguous** t
PC address: 993
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Exception state ====
Type: Informational
Contract: BasicToken
Function name: **ambiguous** t
PC address: 1006
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: solidity_examples/BECToken.sol:28

assert(c >= a)

--------------------

==== Integer Overflow ====
Type: Warning
Contract: StandardToken
Function name: **ambiguous** t
PC address: 2727
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Exception state ====
Type: Informational
Contract: StandardToken
Function name: **ambiguous** t
PC address: 2740
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: solidity_examples/BECToken.sol:28

assert(c >= a)

--------------------

==== Integer Overflow ====
Type: Warning
Contract: PausableToken
Function name: **ambiguous** t
PC address: 4216
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Exception state ====
Type: Informational
Contract: PausableToken
Function name: **ambiguous** t
PC address: 4229
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: solidity_examples/BECToken.sol:28

assert(c >= a)

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: **ambiguous** t
PC address: 5247
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Exception state ====
Type: Informational
Contract: BecToken
Function name: **ambiguous** t
PC address: 5260
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: solidity_examples/BECToken.sol:28

assert(c >= a)

--------------------
```

@JoranHonig what are you getting?

**Comment by JoranHonig** [comment]
I have the following:

```
(venv) > $ ./myth -x ./BECToken.sol:BecToken                                                                                                                                                                                                         [±master ●●●▾]
==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: _function_0x83f12fec
PC address: 1053
A possible integer overflow exists in the function `_function_0x83f12fec`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: ./BECToken.sol:255

function batchTransfer(address[] _receivers, uint256 _value) public whenNotPaused  returns (bool) {
    uint cnt = _receivers.length;
    uint256 amount = uint256(cnt) * _value;

    require(cnt > 0 && cnt <= 20);
    require(_value > 0 );

    balances[msg.sender] = balances[msg.sender].sub(amount);
    for (uint i = 0; i < cnt; i++) {
        balances[_receivers[i]] = balances[_receivers[i]].add(_value);
        Transfer(msg.sender, _receivers[i], _value);
    }
    return true;
  }

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: _function_0x83f12fec
PC address: 2323
A possible integer overflow exists in the function `_function_0x83f12fec`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: ./BECToken.sol:257

uint256(cnt) * _value

--------------------

==== Exception state ====
Type: Informational
Contract: BecToken
Function name: _function_0x83f12fec
PC address: 4830
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: ./BECToken.sol:22

assert(b <= a)

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: transferFrom(address,address,uint256)
PC address: 4848
A possible integer overflow exists in the function `transferFrom(address,address,uint256)`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: ./BECToken.sol:27

a + b

--------------------

==== Exception state ====
Type: Informational
Contract: BecToken
Function name: transferFrom(address,address,uint256)
PC address: 4861
A reachable exception (opcode 0xfe) has been detected. This can be caused by type errors, division by zero, out-of-bounds array access, or assert violations. This is acceptable in most situations. Note however that `assert()` should only be used to check invariants. Use `require()` for regular input checking. 
--------------------
In file: ./BECToken.sol:28

assert(c >= a)

--------------------

```

**Comment by b-mueller** [comment]
Works now! The only annoying thing is that it still says `function name: **ambiguous** t`, but there's already #369 and #377 for that. 

```
$ ./myth -minteger -x solidity_examples/BECToken.sol 
==== Integer Overflow ====
Type: Warning
Contract: PausableToken
Function name: **ambiguous** b
PC address: 659
A possible integer overflow exists in the function `**ambiguous** b`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:255

function batchTransfer(address[] _receivers, uint256 _value) public whenNotPaused returns (bool) {
    uint cnt = _receivers.length;
    uint256 amount = uint256(cnt) * _value;
    require(cnt > 0 && cnt <= 20);
    require(_value > 0 && balances[msg.sender] >= amount);

    balances[msg.sender] = balances[msg.sender].sub(amount);
    for (uint i = 0; i < cnt; i++) {
        balances[_receivers[i]] = balances[_receivers[i]].add(_value);
        Transfer(msg.sender, _receivers[i], _value);
    }
    return true;
  }

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BasicToken
Function name: **ambiguous** t
PC address: 993
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: **ambiguous** b
PC address: 1053
A possible integer overflow exists in the function `**ambiguous** b`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:255

function batchTransfer(address[] _receivers, uint256 _value) public whenNotPaused returns (bool) {
    uint cnt = _receivers.length;
    uint256 amount = uint256(cnt) * _value;
    require(cnt > 0 && cnt <= 20);
    require(_value > 0 && balances[msg.sender] >= amount);

    balances[msg.sender] = balances[msg.sender].sub(amount);
    for (uint i = 0; i < cnt; i++) {
        balances[_receivers[i]] = balances[_receivers[i]].add(_value);
        Transfer(msg.sender, _receivers[i], _value);
    }
    return true;
  }

--------------------

==== Integer Overflow ====
Type: Warning
Contract: PausableToken
Function name: **ambiguous** b
PC address: 1587
A possible integer overflow exists in the function `**ambiguous** b`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:257

uint256(cnt) * _value

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: **ambiguous** b
PC address: 2460
A possible integer overflow exists in the function `**ambiguous** b`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:257

uint256(cnt) * _value

--------------------

==== Integer Overflow ====
Type: Warning
Contract: StandardToken
Function name: **ambiguous** t
PC address: 2727
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Integer Overflow ====
Type: Warning
Contract: PausableToken
Function name: **ambiguous** t
PC address: 4216
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------

==== Integer Overflow ====
Type: Warning
Contract: BecToken
Function name: **ambiguous** t
PC address: 5247
A possible integer overflow exists in the function `**ambiguous** t`.
The addition or multiplication may result in a value higher than the maximum representable integer.
--------------------
In file: solidity_examples/BECToken.sol:27

a + b

--------------------
```

File: mythril/laser/ethereum/instructions.py [hunk 1]
```diff
@@ -778,20 +773,19 @@ def jumpi_(self, global_state):
...
         # False case
         negated = Not(condition) if type(condition) == BoolRef else condition == 0
-        sat = not is_false(simplify(negated)) if type(condi) == BoolRef else not negated
 
-        if sat:
+        if (type(negated) == bool and negated) or (type(negated) == BoolRef and not is_false(simplify(negated))):
             new_state = copy(global_state)
             new_state.mstate.depth += 1
-            new_state.mstate.constraints.append(negated)
+            new_state.mstate.constraints.append(simplify(negated))
```
Comment (norhh) [inline mythril/laser/ethereum/instructions.py]: we can't simplify a bool right?

File: mythril/laser/ethereum/instructions.py [hunk 2]
```diff
@@ -778,20 +773,19 @@ def jumpi_(self, global_state):
                 new_state = copy(global_state)
                 new_state.mstate.pc = index
                 new_state.mstate.depth += 1
-                new_state.mstate.constraints.append(condi)
+                new_state.mstate.constraints.append(simplify(condi))
```
Comment (norhh) [inline mythril/laser/ethereum/instructions.py]: bool can't be simplified right?

**Pull request merged**